### ISSUE TITLE
C++ compilation support?

### ISSUE THREAD

**Issue by brandonros** [boazsegev/facil.io#142]
I know it's a C project, not C++.

```
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:58:13: warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
  route_add("/json", on_request_json);
            ^
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:59:13: warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
  route_add("/plaintext", on_request_plain_text);
            ^
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:81:24: error: non-constant-expression cannot be narrowed from type 'int' to 'int16_t' (aka 'short') in initializer list [-Wc++11-narrowing]
  fio_start(.threads = fio_cli_get_i("-t"), .workers = fio_cli_get_i("-w"));
                       ^~~~~~~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:954:58: note: expanded from macro 'fio_start'
#define fio_start(...) fio_start((struct fio_start_args){__VA_ARGS__})
                                                         ^~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:81:24: note: insert an explicit cast to silence this issue
  fio_start(.threads = fio_cli_get_i("-t"), .workers = fio_cli_get_i("-w"));
                       ^~~~~~~~~~~~~~~~~~~
                       static_cast<int16_t>(
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:954:58: note: expanded from macro 'fio_start'
#define fio_start(...) fio_start((struct fio_start_args){__VA_ARGS__})
                                                         ^~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:81:56: error: non-constant-expression cannot be narrowed from type 'int' to 'int16_t' (aka 'short') in initializer list [-Wc++11-narrowing]
  fio_start(.threads = fio_cli_get_i("-t"), .workers = fio_cli_get_i("-w"));
                                                       ^~~~~~~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:954:58: note: expanded from macro 'fio_start'
#define fio_start(...) fio_start((struct fio_start_args){__VA_ARGS__})
                                                         ^~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:81:56: note: insert an explicit cast to silence this issue
  fio_start(.threads = fio_cli_get_i("-t"), .workers = fio_cli_get_i("-w"));
                                                       ^~~~~~~~~~~~~~~~~~~
                                                       static_cast<int16_t>(
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:954:58: note: expanded from macro 'fio_start'
#define fio_start(...) fio_start((struct fio_start_args){__VA_ARGS__})
                                                         ^~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:94:67: warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
  http_set_header(h, HTTP_HEADER_CONTENT_TYPE, http_mimetype_find("json", 4));
                                                                  ^
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:108:67: warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
  http_set_header(h, HTTP_HEADER_CONTENT_TYPE, http_mimetype_find("txt", 3));
                                                                  ^
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:109:3: error: no matching function for call to 'http_send_body'
  http_send_body(h, "Hello, World!", 13);
  ^~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/http/http.h:197:5: note: candidate function not viable: no known conversion from 'const char [14]' to 'void *' for 2nd argument
int http_send_body(http_s *h, void *data, uintptr_t length);
    ^
/Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:149:18: warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
      database = "localhost";
                 ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:3768:14: error: cannot initialize a variable of type 'fio_str_s *' with an rvalue of type 'void *'
  fio_str_s *str = FIO_MALLOC(sizeof(*str));
             ^     ~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:3769:3: error: no matching literal operator for call to 'operator""__FILE__' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
  FIO_ASSERT_ALLOC(str);
  ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:498:45: note: expanded from macro 'FIO_ASSERT_ALLOC'
    FIO_LOG_FATAL("memory allocation error "__FILE__                           \
                                            ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:3868:14: error: assigning to 'char *' from incompatible type 'void *'
    i.data = tmp;
             ^~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:3878:16: error: assigning to 'char *' from incompatible type 'void *'
      i.data = tmp;
               ^~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4024:5: error: no matching literal operator for call to 'operator""__FILE__' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
    FIO_ASSERT_ALLOC(tmp);
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:498:45: note: expanded from macro 'FIO_ASSERT_ALLOC'
    FIO_LOG_FATAL("memory allocation error "__FILE__                           \
                                            ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4027:5: error: no matching literal operator for call to 'operator""__FILE__' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
    FIO_ASSERT_ALLOC(tmp);
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:498:45: note: expanded from macro 'FIO_ASSERT_ALLOC'
    FIO_LOG_FATAL("memory allocation error "__FILE__                           \
                                            ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4042:3: error: no matching literal operator for call to 'operator""__FILE__' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
  FIO_ASSERT_ALLOC(tmp);
  ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:498:45: note: expanded from macro 'FIO_ASSERT_ALLOC'
    FIO_LOG_FATAL("memory allocation error "__FILE__                           \
                                            ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4077:9: error: assigning to 'char *' from incompatible type 'void *'
  tmp = fio_realloc(s->data, s->len + 1);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4078:3: error: no matching literal operator for call to 'operator""__FILE__' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
  FIO_ASSERT_ALLOC(tmp);
  ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:498:45: note: expanded from macro 'FIO_ASSERT_ALLOC'
    FIO_LOG_FATAL("memory allocation error "__FILE__                           \
                                            ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4230:5: error: cannot jump from this goto statement to its label
    goto at_end;
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4234:15: note: jump bypasses variable initialization
  char *const end = state.data + state.len;
              ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4233:9: note: jump bypasses variable initialization
  char *p = state.data;
        ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4232:11: note: jump bypasses variable initialization
  int32_t c = 0;
          ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4228:5: error: cannot jump from this goto statement to its label
    goto error;
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4234:15: note: jump bypasses variable initialization
  char *const end = state.data + state.len;
              ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4233:9: note: jump bypasses variable initialization
  char *p = state.data;
        ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4232:11: note: jump bypasses variable initialization
  int32_t c = 0;
          ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4340:5: error: cannot jump from this goto statement to its label
    goto zero;
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4342:12: note: jump bypasses variable initialization
  uint64_t l = 0;
           ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4496:14: error: assigning to 'char *' from incompatible type 'void *'
      path = FIO_MALLOC(path_len + 1);
             ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:3097:26: note: expanded from macro 'FIO_MALLOC'
#define FIO_MALLOC(size) fio_malloc((size))
                         ^~~~~~~~~~~~~~~~~~
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4497:7: error: no matching literal operator for call to 'operator""__FILE__' with arguments of types 'const char *' and 'unsigned long', and no matching literal operator template
      FIO_ASSERT_ALLOC(path);
      ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:498:45: note: expanded from macro 'FIO_ASSERT_ALLOC'
    FIO_LOG_FATAL("memory allocation error "__FILE__                           \
                                            ^
In file included from /Users/brandonros/Desktop/j2534-http/vehicle/j2534-http-vehicle.cpp:168:
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4516:5: error: cannot jump from this goto statement to its label
    goto finish;
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4527:16: note: jump bypasses variable initialization
  const size_t org_len = fio_str_len(s);
               ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4511:5: error: cannot jump from this goto statement to its label
    goto finish;
    ^
/Users/brandonros/Desktop/j2534-http/deps/facil.io/lib/facil/fio.h:4527:16: note: jump bypasses variable initialization
  const size_t org_len = fio_str_len(s);
               ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
5 warnings and 20 errors generated.
make[2]: *** [CMakeFiles/j2534-http-vehicle-x64.dir/vehicle/j2534-http-vehicle.cpp.o] Error 1
make[1]: *** [CMakeFiles/j2534-http-vehicle-x64.dir/all] Error 2
make: *** [all] Error 2

```

**Comment by boazsegev** [comment]
Hi @brandonros ,

Thanks for opening the issue, I know it would be nice if facil.io could compile to C++.

I am working on making it as close to possible as I can for the future 0.8.x version... but I won't back port C++ compatibility fixes to the 0.7.x releases.

One of my main issues with C++ compilation is actually the API...

C99 allows to [use the preprocessor together with a `struct` in order to emulate "named arguments"](https://stackoverflow.com/a/5947763/4025095), something that the facil.io API leans on heavily. This allows for a very flexible API and simplifies complex function calls.

However, [C++ does not allow for variadic macros nor out-of-order designated initializers / compound literals](https://pdimov.github.io/blog/2020/09/07/named-parameters-in-c20/), making it impossible to use named macros without GNU extensions (if at all). C++20 seems a step forwards, but there are still issues when trying to bridge the differences.

I am making slow progress on C++ compilation support for the 0.8.x release (which I have been working on for a few years and seem to constantly re-write). Right now the [facil.io C STL](https://github.com/facil-io/cstl), which provides core functionality for the upcoming 0.8.x version, should compile even when used as a C++ header library... but unless there will bridge these two differences between C and C++ (variadic macros and designated initializers) we might have an issue with implementing a comfortable API.

Anyway, thanks for the thought.

Cheers.
   Bo.

**Comment by boazsegev** [comment]
P.S.

AFAIK, there is nothing preventing you from adding a C compilation unit to your project (compiling a C file) and calling the C functions from C++... although it's sub-optimal, it allows you to use facil.io within a C++ project.

B.
